File: SupervisedLearning/FFNNRuby/ai/neuron.rb
class Neuron
  attr_accessor :weights, :output

  def initialize(num_inputs)
    @weights = Array.new(num_inputs) { rand(-1.0..1.0) }
    @num_inputs = num_inputs
  end

  def activate(inputs, bias)
    if @num_inputs != inputs.length then
      abort("Wrong number of inputs")
    end

    sum = bias
    inputs.each_with_index do |input, i|
      sum += input * weights[i]
    end

    @output = sigmoid(sum)
    @output
  end


  def calculate_error(deltas)
    sum = 0.0
    @weights.each do |weight|
      deltas.each do |delta|
        sum += weight * delta
      end
    end

    return sigmoidDerivative(@output) * sum
  end

  # Update the weights of the neuron during backpropagation
  def update_weights(learning_rate, delta)
    @weights = @weights.each_with_index.map do |weight, index|
      weight + learning_rate * delta * @output
    end
  end

end

File: golibs/elman-network/elman/neuron.go
package elman

import (
	"math/rand"

	"github.com/mtricolici/ai-study-2023/golibs/elman-network/emath"
)

type Neuron struct {
	Weights []float64
	Bias    float64
}

func NewNeuron(numInputs int) *Neuron {
	weights := make([]float64, numInputs)
	for i := range weights {
		weights[i] = rand.NormFloat64()
	}
	bias := rand.NormFloat64()
	return &Neuron{weights, bias}
}

func (n *Neuron) Activate(inputs []float64) float64 {
	sum := n.Bias
	for i, input := range inputs {
		sum += input * n.Weights[i]
	}
	return emath.Sigmoid(sum)
}
